web-registry/src/features/ecocredit/CreateBatchBySteps/useCreateBatch.ts

```diff
@@ -0,0 +1,188 @@
+import {
+  generateIri,
+  IriFromMetadataSuccess,
+  stringToUint8Array,
+} from '../../../lib/metadata-graph';
+
+import { CreateBatchFormValues } from './CreateBatchMultiStepForm/CreateBatchMultiStepForm';
+
+// TODO
+// `projectLocation` is hardcoded to 'US' in MsgCreateBatch.fromPartial
+// This case has not been implemented because the data source changes with the update to
+// ecocredits v1, where the information is already part of the project entity in the ledger.
+
+// TODO
+// Right now, just case "C01" (aka. VCS)
+
+function prepareMetadata(
+  partialMetadata: Partial<VCSBatchMetadataLD>,
+): VCSBatchMetadataLD | undefined {
+  const projectId = partialMetadata['regen:vcsProjectId'];
+  const retirementSerialNumber =
+    partialMetadata['regen:vcsRetirementSerialNumber'];
+
+  const additionalCertifications =
+    partialMetadata['regen:additionalCertifications'];
+
+  if (!projectId || !retirementSerialNumber) return;
+
+  const metadata: VCSBatchMetadataLD = {
+    '@context': {
+      schema: 'http://schema.org/',
+      regen: 'http://regen.network/',
+      'regen:additionalCertifications': { '@container': '@list' },
+    },
+    '@type': 'regen:C01-CreditBatch',
+    'regen:vcsProjectId': projectId,
+    'regen:vcsRetirementSerialNumber': retirementSerialNumber,
+    'regen:additionalCertifications': additionalCertifications || [],
+  };
+
+  return metadata;
+}
+
+async function prepareMsg(
+  issuer: string,
+  data: CreateBatchFormValues,
+): Promise<Partial<MsgCreateBatch> | undefined> {
+  // First, complete the metadata
+  const metadata: VCSBatchMetadataLD | undefined = prepareMetadata(
+    data.metadata as Partial<VCSBatchMetadataLD>,
+  );
+  if (!metadata) return;
+
+  // then, generate the offchain iri from the metadata
+  let iriResponse: IriFromMetadataSuccess<VCSBatchMetadataLD> | undefined;
+
+  try {
+    iriResponse = await generateIri(metadata);
+    if (!iriResponse) return;
+  } catch (err) {
+    throw new Error(err as string);
+  }
+
+  // finally, build de Msg DTO
+  const issuance: MsgCreateBatch_BatchIssuance[] = data.recipients.map(
+    recipient => {
+      let issuanceRecipient: Partial<MsgCreateBatch_BatchIssuance> = {
+        recipient: recipient.recipient,
+        tradableAmount: recipient.tradableAmount.toString(),
+        retiredAmount: '0',
+      };
+      if (recipient.withRetire && recipient.retiredAmount > 0) {
+        issuanceRecipient.retiredAmount = recipient.retiredAmount.toString();
+        issuanceRecipient.retirementLocation = recipient.retirementLocation;
+      }
+      return issuanceRecipient as MsgCreateBatch_BatchIssuance;
+    },
+  );
+
+  return MsgCreateBatch.fromPartial({
+    issuer,
+    classId: data.classId,
+    issuance: issuance,
+    metadata: stringToUint8Array(iriResponse.iri),
+    startDate: new Date(data.startDate as Date),
+    endDate: new Date(data.endDate as Date),
+    // TODO - Hardcoded projectLocation
```

Comment:
please also reference this issue since the projectLocation won't be needed anymore starting from v4.0 https://github.com/regen-network/regen-registry/issues/968